boringtun/src/device/mod.rs

```diff
@@ -565,46 +559,50 @@ impl Device {
     }
 
     /// Read from UDP socket, decapsulate, write to tunnel device
-    async fn handle_incoming(device: Weak<RwLock<Self>>, udp: Arc<UdpSocket>) -> Result<(), Error> {
+    async fn handle_incoming(
+        device: Weak<RwLock<Self>>,
+        udp: Arc<<T::UdpTransportFactory as UdpTransportFactory>::Transport>,
+    ) -> Result<(), Error> {
         let mut dst_buf = [0u8; MAX_UDP_SIZE];
 
-        let (buf_tx, mut buf_rx) = mpsc::unbounded_channel::<Box<[u8; 4096]>>();
+        let (buf_tx, mut buf_rx) = mpsc::unbounded_channel::<Box<_>>();
 
         let (tun, private_key, public_key, rate_limiter) = {
             let Some(device) = device.upgrade() else {
                 return Ok(());
             };
             let device = device.read().await;
 
-            let tun = device.tun.clone();
+            let tun = device.tun_tx.clone();
             let (private_key, public_key) = device.key_pair.clone().expect("Key not set");
             let rate_limiter = device.rate_limiter.clone().unwrap();
             (tun, private_key, public_key, rate_limiter)
         };
 
-        let (dec_tx, mut dec_rx) = mpsc::channel::<(PacketBuf, SocketAddr)>(5000);
-        let (send_udp_tx, mut send_udp_rx) = mpsc::channel::<(PacketBuf, SocketAddr)>(5000);
-        let (send_tun_tx, mut send_tun_rx) = mpsc::channel::<PacketBuf>(5000);
+        let (dec_tx, mut dec_rx) = mpsc::channel::<(PacketBuf, SocketAddr)>(MAX_PACKET_BUFS);
+        let (send_udp_tx, mut send_udp_rx) =
+            mpsc::channel::<(PacketBuf, SocketAddr)>(MAX_PACKET_BUFS);
+        let (send_tun_tx, mut send_tun_rx) = mpsc::channel::<PacketBuf>(MAX_PACKET_BUFS);
 
         let udp_send = udp.clone();
         let buf_tx_udp = buf_tx.clone();
-        let mut send_task_udp = AbortOnDrop(tokio::task::spawn(async move {
+        let mut send_task_udp = Task::spawn("send_udp", async move {
             while let Some((packet_buf, addr)) = send_udp_rx.recv().await {
-                let _: Result<_, _> = udp_send.send_to(packet_buf.packet(), &addr).await;
+                let _: Result<_, _> = udp_send.send_to(packet_buf.packet(), addr).await;
                 let _ = buf_tx_udp.send(packet_buf.buf);
             }
-        }));
+        });
         let buf_tx_tun = buf_tx.clone();
-        let mut send_task_tun = AbortOnDrop(tokio::task::spawn(async move {
+        let mut send_task_tun = Task::spawn("send_tun", async move {
             while let Some(packet_buf) = send_tun_rx.recv().await {
                 let _: Result<_, _> = tun.send(packet_buf.packet()).await;
                 let _ = buf_tx_tun.send(packet_buf.buf);
             }
-        }));
+        });
 
         let buf_tx_encapsulate = buf_tx.clone();
         let udp_send_flush = udp.clone(); // TODO: do we need this?
-        let mut encapsulate_task = AbortOnDrop(tokio::task::spawn(async move {
+        let mut encapsulate_task = Task::spawn("encapsulate", async move {
```

Comment:
Should this task now be named something else than encapsulate as it decapsulates incoming packets? The task in `handle_outgoing` is also named "encapsulate" which is confusing.

Reply:
Yes, this is incorrect. Fixed.